### PR TITLE
Basic support for option types

### PR DESCRIPTION
**Pull request by wytseoortwijn**
Adds some basic support for option types.

### CODE REVIEW

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (ArquintL) [inline src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala:16]: I like the var name in OptionToSeqImpl better:
```
  /**
     * Determines whether the Viper domain
     * should be generated upon finalisation.
     */
   private var generateDomain : Boolean = false
```

**Review by ArquintL — APPROVED** [review]

File: src/main/scala/viper/gobra/frontend/Parser.scala
Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/frontend/Parser.scala:1115]: No, you are right in that "get" is much more standard and intuitive.

However, I didn't want to make "get" a reserved keyword, since it is probably used a lot in 'real-world' Go programs. So I opted for "option" instead, since it already is a reserved keyword.

Do you think it is better to use "get"?

Comment (Felalolf) [reply src/main/scala/viper/gobra/frontend/Parser.scala:1115]: Some langauges use "the", I believe.

Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/Parser.scala:1115]: I agree that it could be pretty annoying to have `get` as a reserved keyword. Do we have a general strategy so far? Simply don't care and force programmers to respect all reserved keywords or go for some special prefix to decrease likelihood (or even better use an invalid prefix in go)?

Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/frontend/Parser.scala:1115]: I don't think there is a general strategy at the moment.

One possibility is to make `gobra` a reserved keyword, and use something like `gobra:get`.

**Review by jcp19 — APPROVED** [review]

File: src/main/scala/viper/gobra/frontend/Parser.scala
Comment (jcp19) [reply src/main/scala/viper/gobra/frontend/Parser.scala:1115]: That approach seems reasonable, if we ever need to add extra operations to the language we can always use the `gobra` keyword to prefix them

Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/Parser.scala:1115]: I like this idea as well. Another extension of that would be to write a "fake" import that would allow you to specify the qualifier (`gobra`, `.` or any other identifier)

Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/frontend/Parser.scala:1115]: Yes, I like the "fake import" idea.

Comment (Felalolf) [reply src/main/scala/viper/gobra/frontend/Parser.scala:1115]: Me too.

Comment (Felalolf) [reply src/main/scala/viper/gobra/frontend/Parser.scala:1115]: There is a small problem there. Since Go does not support polymorphism, so the import does really have to be fake, since the functions we provide could not be defined in Go.

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala]: I agree. Just updated it.

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (Felalolf) [inline src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala:79]: Please, add a trigger.

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (Felalolf) [inline src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala:110]: Please, add a trigger.

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (Felalolf) [inline src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala:135]: This is the last time that I will write it, but all quantified expressions should get a trigger. When you want to use the default trigger, then instantiate it explicitly. We should always know what our triggers are.

File: src/main/scala/viper/gobra/translator/implementations/components/OptionToSeqImpl.scala
Comment (Felalolf) [inline src/main/scala/viper/gobra/translator/implementations/components/OptionToSeqImpl.scala:43]: The axiomatization seems weird to me. I would expect something along the lines of:
`opt2seq(optnone()) == Seq()  && forall e :: opt2seq(optsome(e)) == Seq(e)`.
A separate axiom should specify that an option is either some or none. I think we had some previous work on how ADT are encoded into Viper.

Comment (Felalolf) [reply src/main/scala/viper/gobra/translator/implementations/components/OptionToSeqImpl.scala:43]: I think that we usually use some "tag" utility function to distinguish between the elements.

**Review by Felalolf — CHANGES_REQUESTED** [review]
Triggers are missing, I have some questions regarding the encoding, and the files GobraStrategy.scala and Nodes.scala have to be adapted. Lastly, the option getter should be changed from "option" to either "the" or "get". Regarding the proposed handling of name clashes, someone should create an issue.

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala]: That sounds reasonable. I updated (most of) the axioms.

File: src/main/scala/viper/gobra/translator/implementations/components/OptionToSeqImpl.scala
Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/translator/implementations/components/OptionToSeqImpl.scala]: I updated the axiom according to your suggestion. Regarding "tag" functions; do you have an example or pointer?

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala]: Done

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala:70]: Do you have a suggestion for a trigger for
```
axiom {
  forall o : Option[T] :: o == optnone() || exists e : T :: o == optsome(e)
}
```
or otherwise an alternative axiom?

File: src/main/scala/viper/gobra/frontend/Desugar.scala
Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/frontend/Desugar.scala]: Fixed

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala]: Nevermind, this axiom is superfluous

File: src/main/scala/viper/gobra/translator/implementations/components/OptionToSeqImpl.scala
Comment (Felalolf) [reply src/main/scala/viper/gobra/translator/implementations/components/OptionToSeqImpl.scala]: I will send you a mail with two sources in a minute.

**Comment by Felalolf** [comment]
Please write when I should have a look at the PR again.

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (Felalolf) [inline src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala:81]: This axiom is unsound (incombination with other axioms). 
My - longer - explanation:
1) The axiom implies that get is injective without any restriction
2) If get is complete on type option, then get is not injective, because `get(none) == get(some(get(none)))` and `none != some(...)`

I also double checked with Martin and his - shorter - explanation is: substitute `o` with `none` and `e` with `get(none)`, then the axiom yields `none == some(get(none))` which again is invalid.

File: src/main/scala/viper/gobra/frontend/Parser.scala
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/Parser.scala:1115]: can you change the name before it is merged in? You could use the, get, option-get, etc

**Review by Felalolf — CHANGES_REQUESTED** [review]

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala]: Yes, that's indeed true; thanks for spotting that!

I updated the axioms to be similar to the ADT encoding example on http://viper.ethz.ch/examples/encoding-adts.html.

(I could make it slightly less elaborate if that's better, by just using a simple `tag` function to replace the constructor type business.)

File: src/main/scala/viper/gobra/frontend/Parser.scala
Comment (wytseoortwijn) [reply src/main/scala/viper/gobra/frontend/Parser.scala]: Yes, fixed it. It now uses "get".

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (Felalolf) [inline src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala:328]: The domain is indeed quite large now.
How about something like the following:

```
domain Option[T] {

  function some(e: T): Option[T]
  function unit(): Option[T]
  function get(o: Option[T]): T
  function isUnit(o: Option[T]): Bool

  axiom {
    forall e: T :: { some(e) } get(some(e)) == e && isUnit(some(e)) == false
  }

  axiom {
    forall o: Option[T] :: { get(o) } isUnit(o) == false ==> o == some(get(o))
  }

  axiom {
    isUnit((unit(): Option[T])) == true
  }

  axiom {
    forall o: Option[T] :: { isUnit(o) } o == unit() || exists e: T :: {some(e)} o == some(e)
  }
}
```

With some tests:
```
domain Dummy[T] {
  function dummy(x: T): Bool
  axiom {
    forall x: T :: { dummy(x) } dummy(x)
  }
}


method test1() {
  var x: Option[Int] := some(4)
  assert x != unit()
  assert get(x) == 4
}

method test2(x: Option[Int]) {
  if (x != unit()) {
    assert dummy(get(x)) // hard to avoid
    assert exists e: Int :: some(e) == x
  }
}
```

File: src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala
Comment (Felalolf) [inline src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala:293]: I was not even aware of this viper feature. Why do you not just substitute values such as 0 and 1?

**Review by Felalolf — APPROVED** [review]

**Pull request merged**
